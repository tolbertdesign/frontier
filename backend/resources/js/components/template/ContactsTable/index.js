import ContactsTable from './ContactsTable.vue'

export default ContactsTable
