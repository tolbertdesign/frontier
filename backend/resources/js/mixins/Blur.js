export default {
  methods: {
    blur () {
      document.getElementById('app').style.filter = 'blur(4px)'
    },
    unBlur () {
      document.getElementById('app').style.filter = 'none'
    },
  },
}
