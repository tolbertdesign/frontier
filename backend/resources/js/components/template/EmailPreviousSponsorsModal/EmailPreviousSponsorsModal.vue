<template>
  <div class="m-4 rounded-xl overflow-hidden previous-sponsor-modal">
    <div class="card bg-blue">
      <div
        v-if="isSending"
        class="card-content pt-12 pb-12 px-4 text-white text-center"
      >
        <SendEnvelope />
        <p class="font-bold text-xl">
          {{ lang.email_send_to }}:
        </p>
        <p class="font-semibold pt-4">
          {{ previousSponsors.length }} <span>{{ lang.contact }}</span>
        </p>
      </div>
      <div
        v-else
        class="card-content pt-16 px-4 text-white text-center"
      >
        <SendEnvelope />
        <p class="font-medium text-lg">
          {{ lang.about_to_email_previous }}
        </p>
      </div>
      <div
        v-if="!isSending"
        class="card-footer border-t-0"
      >
        <div class="card-footer-item">
          <button
            class="w-24 sm:w-32 button inline-block mb-4 mx-2 is-danger is-rounded shadow font-bold"
            @click="close"
          >
            {{ lang.cancel }}
          </button>
          <button
            class="w-24 sm:w-32 button inline-block mb-4 mx-2 is-secondary is-rounded is-inverted shadow font-bold"
            @click="sendEmail"
          >
            {{ lang.send }}
          </button>
        </div>
      </div>
    </div>
  </div>
</template>

<script>
import Blur from '@/mixins/Blur'

export default {
  name: 'EmailPreviousSponsorsModal',
  mixins: [Blur],
  props: {
    previousSponsors: {
      type: Array,
      default: null,
    },
    participantUserId: {
      type: Number,
      default: null,
    },
  },
  data () {
    return {
      isSending: false,
    }
  },
  computed: {
    lang () {
      return this.$store.state.lang
    },
  },
  methods: {
    sendEmail () {
      const sponsors = this.previousSponsors

      if (sponsors.length === 0) {
        this.close()
        return
      }

      const sponsorIds = sponsors.map(sponsor => sponsor.id)

      axios.post('/v3/api/previous-contact-enroll', {
        participantUserId: this.participantUserId,
        sponsorIds: sponsorIds,
      }).then(response => {
        this.isSending = true
        this.$emit('previousSponsorsEmailed', response.data.emailed, response.data.skipped)

        setTimeout(() => {
          this.close()
        }, 3000)
      }).catch(error => {
        console.error(error)
        setTimeout(() => {
          this.close()
        }, 3000)
      })
    },
    close () {
      this.$emit('close')
      this.unBlur()
    },
  },
}
</script>
