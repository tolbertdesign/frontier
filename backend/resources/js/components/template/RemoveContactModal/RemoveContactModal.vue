<template>
  <div class="m-4 rounded-xl overflow-hidden">
    <div
      v-if="isSending"
      class="card bg-blue w-full"
    >
      <div
        v-if="isSending"
        class="card-content py-16 text-white text-center"
      >
        <SendEnvelope />
        <p class="font-bold text-xl">
          {{ lang.contact_removed }}
        </p>
      </div>
    </div>
    <div
      v-else
      class="card bg-blue w-full"
    >
      <header class="card-header p-4 shadow-none">
        <p class="flex justify-between items-center text-center modal-card-title text-white">
          <span style="width: 24px; height: 48px;" />
          <span class="flex-1 text-center text-2xl font-bold">Remove Contact</span>
          <a
            href="#"
            class="text-white self-start"
            style="width: 24px; height: 48px;"
            @click.prevent="closeModal"
          >
            <span class="icon">
              <FontAwesomeIcon
                :icon="['fas', 'times']"
                size="xs"
              />
            </span>
          </a>
        </p>
      </header>
      <div
        class="card-content text-lg text-white text-center"
      >
        <p class="px-2">
          Removing a contact will remove them from this list for the current email campaign.
        </p>
      </div>
      <div class="card-footer border-t-0">
        <div class="card-footer-item flex-col mb-4">
          <button
            class="px-8 rounded-full button inline-block is-danger shadow-md font-bold"
            @click="removeContact"
          >
            Remove Contact
          </button>
        </div>
      </div>

    </div>

  </div>
</template>

<script>
import { mapState } from 'vuex'
import Blur from '@/mixins/Blur'

export default {
  name: 'RemoveContactModal',
  mixins: [Blur],
  props: {
    contact: {
      type: Object,
      default: null,
    },
  },
  data: function () {
    return {
      isSending: false,
    }
  },
  computed: {
    ...mapState(['lang']),
  },
  methods: {
    removeContact () {
      // call action to remove contact
      this.isSending = true
      this.$store
        .dispatch(
          'removeContact',
          {
            contact: this.contact,
          },
        )
        .then(() => {
          setTimeout(() => {
            this.closeModal()
            this.isSending = false
          }, 2000)
        })
    },
    closeModal () {
      this.$emit('close')
      this.unBlur()
    },
  },
}
</script>
