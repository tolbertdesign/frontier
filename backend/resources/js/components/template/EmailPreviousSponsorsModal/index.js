import EmailPreviousSponsorsModal from './EmailPreviousSponsorsModal.vue'

export default EmailPreviousSponsorsModal
