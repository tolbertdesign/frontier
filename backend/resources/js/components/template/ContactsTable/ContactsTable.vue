<template>
  <section>
    <div class="contacts-table px-4 pb-4">
      <div class="border border-transparent pl-3 pb-1">
        <div
          v-if="allSelected"
          class="pb-5"
        >
          <div class="text-center">
            <button
              :disabled="noSponsorsSelected"
              class="w-64 button px-10 font-semibold is-rounded is-secondary shadow"
              @click="emailPreviousSponsors"
            >{{ lang.email_previous_sponsors }}</button>
          </div>
        </div>
        <BCheckbox
          v-if="contacts.length"
          :value="allSponsorsSelected"
          type="is-secondary"
          @change.native="toggleSelectAll"
        >
          <span class="inline-block pl-2 sm:pl-3">
            {{ allSponsorsSelected ? lang.unselect_all : lang.select_all }}
          </span>
        </BCheckbox>
      </div>

      <table class="table is-fullwidth border">
        <!-- Show contacts header, if there are any -->
        <thead v-if="contacts.length">
          <tr class="bg-secondary text-white">
            <th colspan="2">
              {{ lang.contact }}
            </th>
            <th>{{ lang.status }}</th>
            <th>
              <FontAwesomeIcon
                v-if="!isEditing"
                :icon="['far', 'user-edit']"
                class="cursor-pointer"
                @click="isEditing = true"
              />
              <button
                v-else
                class="button is-secondary is-inverted is-small"
                @click="isEditing = false"
              >
                <span>Close</span>
                <span class="icon">
                  <FontAwesomeIcon :icon="['fas', 'times']" />
                </span>
              </button>
            </th>
          </tr>
        </thead>

        <!-- Show the contacts, if we have them  -->
        <tbody v-if="contacts.length">
          <tr
            v-for="(contact, index) in getContactsSorted"
            :key="index"
          >
            <td class="align-middle">
              <BCheckbox
                v-if="contact.isSelectable"
                :value="contact.isSelected"
                type="is-secondary"
                class="custom"
                @change.native="selectContact(contact.id)"
              />
            </td>
            <td class="w-full">
              <div class="flex-1 sm:flex items-center justify-between">
                <div class="flex-1">
                  {{ contact.first_name }} {{ contact.last_name }}
                  <br>
                  {{ contact.email }}
                </div>
                <div
                  v-if="contact.isPreviousSponsor"
                  class="italic font-semibold has-text-secondary sm:pr-10"
                >
                  {{ lang.previous_sponsor }}
                </div>
              </div>
            </td>
            <td>{{ translateStatus(contact.status) }}</td>
            <td class="text-center align-middle">
              <FontAwesomeIcon
                v-if="isEditing"
                :icon="['fas', 'trash-alt']"
                @click="removeContact(contact)"
              />
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </section>
</template>

<script>
import RemoveContactModal from '@/components/template/RemoveContactModal'
import EmailPreviousSponsorsModal from '@/components/template/EmailPreviousSponsorsModal'
import { PledgingStatus } from '@/utilities/PledgingStatus'
import Blur from '@/mixins/Blur'

export default {
  name: 'ContactsTable',
  mixins: [Blur],
  props: {
    allSelected: {
      type: Boolean,
      default: true,
    },
    contacts: {
      type: Array,
      default: null,
    },
    participantUserId: {
      type: Number,
      default: null,
    },
  },
  data () {
    return {
      isEditing: false,
    }
  },
  computed: {
    lang () {
      return this.$store.state.lang
    },
    previousSponsors () {
      return this.contacts.filter(contact => contact.isPreviousSponsor)
    },
    previousSelectedSponsors () {
      return this.previousSponsors.filter(contact => contact.isSelected)
    },
    noSponsorsSelected () {
      return this.previousSponsors.filter(contact => contact.isSelected).length === 0
    },
    allSponsorsSelected () {
      const numOfPreviousSponsors = this.previousSponsors.filter(contact => contact.isSelectable).length
      return this.previousSponsors.filter(contact => contact.isSelected).length === numOfPreviousSponsors
    },
    getContactsSorted () {
      const sortedContacts = []

      this.contacts.forEach(function (contact, index) {
        sortedContacts.push(Object.assign({}, contact))
      })

      return sortedContacts.sort(this.sortContact)
    },
  },
  methods: {
    sortContact (contactA, contactB) {
      const statusSortResult = this.sortStatusCompare(contactA.status, contactB.status)

      if (statusSortResult !== 0) {
        return statusSortResult
      }

      if (contactA.first_name && contactB.first_name) {
        const firstNameSortResult = this.sortStringCompare(contactA.first_name, contactB.first_name)

        if (firstNameSortResult !== 0) {
          return firstNameSortResult
        }
      }

      if (contactA.last_name && contactB.last_name) {
        const lastNameSortResult = this.sortStringCompare(contactA.last_name, contactB.last_name)

        if (lastNameSortResult !== 0) {
          return lastNameSortResult
        }
      }

      const emailSortResult = this.sortStringCompare(contactA.email, contactB.email)

      if (emailSortResult !== 0) {
        return emailSortResult
      }

      return -1
    },
    sortStringCompare (stringOne, stringTwo) {
      var firstNameA = stringOne.toLowerCase().trim()
      var firstNameB = stringTwo.toLowerCase().trim()

      if (firstNameA > firstNameB) {
        return 1
      } else if (firstNameA < firstNameB) {
        return -1
      } else {
        return 0
      }
    },
    sortStatusCompare (statusOne, statusTwo) {
      statusOne = statusOne.toLowerCase().trim()
      statusTwo = statusTwo.toLowerCase().trim()

      if (statusOne === statusTwo) {
        return 0
      }

      const statusOrder = Object.keys(PledgingStatus).map(function (key) {
        return PledgingStatus[key].toLowerCase()
      })

      return statusOrder.indexOf(statusOne) - statusOrder.indexOf(statusTwo)
    },
    addContact (contact) {
      this.contacts.push(contact)
    },
    selectContact (contactId) {
      const contact = this.contacts.find(contact => contact.id === contactId)
      contact.isSelected = !contact.isSelected
    },
    toggleSelectAll () {
      const isSelected = !this.allSponsorsSelected
      this.contacts.filter(contact => contact.isSelectable).forEach(contact => {
        contact.isSelected = isSelected
      })
    },
    removeContact (contact) {
      this.$buefy.modal.open({
        parent: this,
        component: RemoveContactModal,
        hasModalCard: true,
        width: 320,
        onCancel: this.unBlur,
        props: {
          contact,
        },
      })
      this.blur()
      this.isEditing = false
    },
    translateStatus (statusText) {
      return this.lang.easy_emailer_status[statusText.toLowerCase()]
    },
    convertPreviousToPotentialSponsors (sponsorIds, status) {
      this.contacts.forEach(contact => {
        if (sponsorIds.indexOf(contact.id) !== -1 && contact.isPreviousSponsor) {
          contact.status = status
          contact.isSelectable = false
          contact.isSelected = false
        }
      })
    },
    emailPreviousSponsors () {
      this.$buefy.modal.open({
        parent: this,
        component: EmailPreviousSponsorsModal,
        hasModalCard: true,
        onCancel: this.unBlur,
        props: {
          previousSponsors: this.previousSelectedSponsors,
          participantUserId: this.participantUserId,
        },
        events: {
          previousSponsorsEmailed: (emailedSponsorIds, skippedSponsorIds) => {
            this.convertPreviousToPotentialSponsors(emailedSponsorIds, PledgingStatus.ACTIVE)
            this.convertPreviousToPotentialSponsors(skippedSponsorIds, PledgingStatus.UNSUBSCRIBED)
          },
        },
      })
      this.blur()
    },
  },
}
</script>

<style>
.table thead td,
.table thead th {
  color: #ffffff;
}
.table td,
.table th {
  vertical-align: middle;
}
.table thead th {
  height: 4em;
}
.b-checkbox.checkbox.custom .control-label {
  padding-left: 0;
}
td.checkbox-cell label.is-disabled {
  display: none;
}
</style>
