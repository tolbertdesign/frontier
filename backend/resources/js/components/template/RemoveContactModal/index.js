import RemoveContactModal from './RemoveContactModal.vue'

export default RemoveContactModal
