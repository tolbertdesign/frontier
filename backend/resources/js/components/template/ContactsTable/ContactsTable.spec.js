import { shallowMount, createLocalVue } from '@vue/test-utils'
import ContactsTable from './ContactsTable.vue'
import RemoveContactModal from '@/components/template/RemoveContactModal'
import EmailPreviousSponsorsModal from '@/components/template/EmailPreviousSponsorsModal'
import PledgingStatus from '@/utilities/PledgingStatus'
import Blur from '@/mixins/Blur'

const localVue = createLocalVue()

const $store = {
  state: {
    lang: {
      email_previous_sponsors: 'email previous sponsors',
      select_all: 'select all',
      unselect_all: 'unselect all',
      contact: 'contact',
      status: 'status',
      previous_sponsor: 'previous sponsor',
      easy_emailer_status: [
        'sent',
        'not sent',
      ],
    },
  },
}
const contacts = []
const allSelected = true

describe('ContactsTable', () => {
  it('Snapshop test for ContactsTable', () => {
    const wrapper = shallowMount(ContactsTable, {
      localVue,
      mixins: {
        Blur,
      },
      components: {
        RemoveContactModal,
        EmailPreviousSponsorsModal,
        PledgingStatus,
      },
      mocks: {
        $store,
      },
      propsData: {
        allSelected: allSelected,
        contacts: contacts,
      },
    })
    expect(wrapper).toMatchSnapshot()
  })
})
