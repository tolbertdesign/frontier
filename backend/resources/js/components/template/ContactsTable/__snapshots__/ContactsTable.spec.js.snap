// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactsTable Snapshop test for ContactsTable 1`] = `
<section>
  <div class="contacts-table px-4 pb-4">
    <div class="border border-transparent pl-3 pb-1">
      <div class="pb-5">
        <div class="text-center"><button disabled="disabled" class="w-64 button px-10 font-semibold is-rounded is-secondary shadow">email previous sponsors</button></div>
      </div>
      <!---->
    </div>
    <table class="table is-fullwidth border">
      <!---->
      <!---->
    </table>
  </div>
</section>
`;
